File: catkin_ws/src/learning_topic/src/velocity_publisher.cpp
//
// Created by jyh on 2020/2/23.
//

/**
 * This routine will publish the turtle1/cmd_vel topic, message type is geometry_msgs::Twist
 */

#include <ros/ros.h>
#include <geometry_msgs/Twist.h>

int main(int argc, char **argv)
{
    // ROS init
    ros::init(argc, argv, "velocity_publisher");

    // Create a node handle
    ros::NodeHandle n;

    // Create a Publisher, publish a topic named /turtle1/cmd_vel, message type: geometry_msgs:Twist, queue length 10
    ros::Publisher turtle_vel_pub = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);

    // Set the frequency of the loop
    ros::Rate loop_rate(10);

    int count = 0;
    while (ros::ok())
    {
        // Initialize the message of  geometry_msgs::Twist
        geometry_msgs::Twist vel_msg;
        vel_msg.linear.x = 0.5;
        vel_msg.angular.z = 0.2;

        // publish message
        turtle_vel_pub.publish(vel_msg);
        ROS_INFO("Publsh turtle velocity command[%0.2f m/s, %0.2f rad/s]",
                 vel_msg.linear.x, vel_msg.angular.z);

        // Delay by loop frequency
        loop_rate.sleep();
    }

    return 0;
}

File: catkin_ws/src/learning_launch/scripts/velocity_publisher.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# This routine will publish the turtle1/cmd_vel topic, message type is geometry_msgs::Twist

import rospy
from geometry_msgs.msg import Twist

def velocity_publisher():
  # ROS init
  rospy.init_node('velocity_publisher', anonymous=True)

  # Create a Publisher, publish a topic named /turtle1/cmd_vel, message type: geometry_msgs:Twist, queue length 10
  turtle_vel_pub = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
  publish_frequency = rospy.get_param('~publish_frequency')
  #Set the frequency of the loop
  rate = rospy.Rate(publish_frequency)

  while not rospy.is_shutdown():
    # Initialize the message of  geometry_msgs::Twist
    vel_msg = Twist()
    vel_msg.linear.x = rospy.get_param('~linear_x')
    vel_msg.angular.z = rospy.get_param('~group/angular_z')

    # pulish message
    turtle_vel_pub.publish(vel_msg)
    rospy.loginfo("Publsh turtle velocity command[%0.2f m/s, %0.2f rad/s]",
                  vel_msg.linear.x, vel_msg.angular.z)

    # Delay by cycle frequency
    rate.sleep()

if __name__ == '__main__':
  try:
    velocity_publisher()
  except rospy.ROSInterruptException:
    pass